cmd/downloader/downloader/downloader_grpc_server.go

```diff
@@ -32,49 +33,35 @@ func (s *GrpcServer) Download(ctx context.Context, request *proto_downloader.Dow
 	defer logEvery.Stop()
 
 	torrentClient := s.d.Torrent()
-	mi := &metainfo.MetaInfo{AnnounceList: Trackers}
+	snapDir := s.d.SnapDir()
 	for i, it := range request.Items {
-		if it.TorrentHash == nil { // seed new snapshot
-			if err := BuildTorrentFileIfNeed(it.Path, s.d.SnapDir()); err != nil {
+		var ok bool
+		var err error
+		if it.TorrentHash == nil {
+			// if we dont have the torrent hash then we seed a new snapshot
+			ok, err = seedNewSnapshot(it, torrentClient, snapDir)
+			if err != nil {
 				return nil, err
 			}
 		}
 
-		hash := Proto2InfoHash(it.TorrentHash)
-		if _, ok := torrentClient.Torrent(hash); ok {
-			continue
-		}
-
-		ok, err := AddSegment(it.Path, s.d.SnapDir(), torrentClient)
-		if err != nil {
-			return nil, fmt.Errorf("AddSegment: %w", err)
-		}
 		select {
 		case <-logEvery.C:
-			log.Info("[snpshots] initializing", "files", fmt.Sprintf("%d/%d", i, len(request.Items)))
+			log.Info("[snapshots] initializing", "files", fmt.Sprintf("%d/%d", i, len(request.Items)))
 		default:
 		}
 		if ok {
```

Comment:
Probably can move this inside “ if it.TorrentHash == nil {” and move logs select to beginning of for loop